lib/internal/loader/Loader.js

```diff
@@ -40,105 +40,83 @@ function normalizeReferrerURL(referrer) {
  * the main module and everything in its dependency graph. */
 class Loader {
   constructor(base = getURLStringForCwd()) {
-    if (typeof base !== 'string') {
+    if (typeof base !== 'string')
       throw new errors.TypeError('ERR_INVALID_ARG_TYPE', 'base', 'string');
-    }
 
-    this.moduleMap = new ModuleMap();
     this.base = base;
+
+    // methods which translate input code or other information
+    // into es modules
+    this.translators = translators;
+
+    // registry of loaded modules, akin to `require.cache`
+    this.moduleMap = new ModuleMap();
+
     // The resolver has the signature
     //   (specifier : string, parentURL : string, defaultResolve)
-    //       -> Promise<{ url : string,
-    //                    format: anything in Loader.validFormats }>
+    //       -> Promise<{ url : string, format: string }>
     // where defaultResolve is ModuleRequest.resolve (having the same
     // signature itself).
     // If `.format` on the returned value is 'dynamic', .dynamicInstantiate
     // will be used as described below.
-    this.resolver = ModuleRequest.resolve;
-    // This hook is only called when resolve(...).format is 'dynamic' and has
-    // the signature
+    this._resolve = defaultResolve;
+    // This hook is only called when resolve(...).format is 'dynamic' and
+    // has the signature
     //   (url : string) -> Promise<{ exports: { ... }, execute: function }>
     // Where `exports` is an object whose property names define the exported
     // names of the generated module. `execute` is a function that receives
     // an object with the same keys as `exports`, whose values are get/set
     // functions for the actual exported values.
-    this.dynamicInstantiate = undefined;
-  }
-
-  hook({ resolve = ModuleRequest.resolve, dynamicInstantiate }) {
-    // Use .bind() to avoid giving access to the Loader instance when it is
-    // called as this.resolver(...);
-    this.resolver = resolve.bind(null);
-    this.dynamicInstantiate = dynamicInstantiate;
+    this._dynamicInstantiate = undefined;
   }
 
-  // Typechecking wrapper around .resolver().
   async resolve(specifier, parentURL = this.base) {
-    if (typeof parentURL !== 'string') {
-      throw new errors.TypeError('ERR_INVALID_ARG_TYPE',
-                                 'parentURL', 'string');
-    }
-
-    const { url, format } = await this.resolver(specifier, parentURL,
-                                                ModuleRequest.resolve);
-
-    if (!Loader.validFormats.includes(format)) {
-      throw new errors.TypeError('ERR_INVALID_ARG_TYPE', 'format',
-                                 Loader.validFormats);
-    }
-
-    if (typeof url !== 'string') {
-      throw new errors.TypeError('ERR_INVALID_ARG_TYPE', 'url', 'string');
-    }
-
-    if (format === 'builtin') {
-      return { url: `node:${url}`, format };
-    }
```

Comment:
These protections are contained outside of `_resolve` because when a user hooks the resolve hook, we still make sure to validate the input and output arguments for ease of writing a loader with a resolve hook.

Reply:
is the below good? checking if the format is valid is done in getModuleJob

```diff
   async resolve(specifier, parentURL = this.base) {
-    return this._resolve(specifier, parentURL, defaultResolve);
+    if (typeof parentURL !== 'string')
+      throw new errors.TypeError('ERR_INVALID_ARG_TYPE', 'parentURL', 'string');
+
+    const { url, format } =
+      await this._resolve(specifier, parentURL, defaultResolve);
+
+    if (typeof url !== 'string')
+      throw new errors.TypeError('ERR_INVALID_ARG_TYPE', 'url', 'string');
+
+    if (typeof format !== 'string')
+      throw new errors.TypeError('ERR_INVALID_ARG_TYPE', 'format', 'string');
+
+    if (format === 'builtin')
+      return { url: `node:${url}`, format };
+
+    return { url, format };
   }
```

Reply:
Looks good to me.

Ideally these errors should all provide their parentURL in the error message for a better debugging experience... but improving these cases can be tackled in due course.

Reply:
I think it's also missing the `file://` protocol check here?

Also, you don't seem to have included the `node:` builtin concatenation? Not having this would effectively be altering the loader API to expect `node:` included for builtins, which was an interface we specially didn't want to expose here.

Reply:
file protocol check is only needed if people are using the builtin stuff (which has the check), if they override it with their own resolve/dynamic they needn't be limited by that

Reply:
That would be a new change in this PR. Currently URLs are restricted to file:/// URLs even for dynamic resolves to ensure consistency. //cc @bmeck.